evm/chains/sharding/mainchain_handler/vmc_utils.py

```diff
@@ -0,0 +1,212 @@
+    return _valmgr_code
+
+def get_valmgr_bytecode():
+    global _valmgr_bytecode
+    if not _valmgr_bytecode:
+        _valmgr_bytecode = compiler.compile(get_valmgr_code())
+    return _valmgr_bytecode
+
+def get_valmgr_addr():
+    global _valmgr_addr
+    if not _valmgr_addr:
+        create_valmgr_tx()
+    return _valmgr_addr
+
+def get_valmgr_sender_addr():
+    global _valmgr_sender_addr
+    if not _valmgr_sender_addr:
+        create_valmgr_tx()
+    return _valmgr_sender_addr
+
+def get_valmgr_tx():
+    global _valmgr_tx
+    if not _valmgr_tx:
+        create_valmgr_tx()
+    return _valmgr_tx
+
+def create_valmgr_tx(gasprice=GASPRICE):
+    global _valmgr_sender_addr, _valmgr_addr, _valmgr_tx
+    bytecode = get_valmgr_bytecode()
+    v = 27
+    r = 1000000000000000000000000000000000000000000000000000000000000000000000000000
+    s = 1000000000000000000000000000000000000000000000000000000000000000000000000000
+    tx = SpuriousDragonTransaction(0, gasprice, 3000000, b'', 0, bytecode, v, r, s)
+    valmgr_sender_addr = tx.sender
+    valmgr_addr = eth_utils.to_checksum_address(
+        generate_contract_address(eth_utils.to_canonical_address(valmgr_sender_addr), 0)
+    )
+    _valmgr_sender_addr = eth_utils.to_checksum_address(valmgr_sender_addr)
+    _valmgr_addr = valmgr_addr
+    _valmgr_tx = tx
+
+def mk_initiating_contracts(sender_privkey, sender_starting_nonce, gasprice=GASPRICE):
+    """Make transactions of createing initial contracts
+    Including rlp_decoder, sighasher and validator_manager
+    """
+    o = []
+    nonce = sender_starting_nonce
+    global viper_rlp_decoder_tx, sighasher_tx
+    # the sender gives all senders of the txs money, and append the
+    # money-giving tx with the original tx to the return list
+    for tx in (viper_rlp_decoder_tx, sighasher_tx, get_valmgr_tx()):
+        o.append(
+            SpuriousDragonUnsignedTransaction(
+                nonce,
+                gasprice,
+                500000,
+                tx.sender,
+                tx.gas * tx.gas_price + tx.value,
+                b'',
+            ).as_signed_transaction(sender_privkey)
+        )
+        nonce += 1
+    o += [viper_rlp_decoder_tx, sighasher_tx, get_valmgr_tx()]
+    return o
+
+def get_func_abi(func_name, contract_abi):
+    for func_abi in contract_abi:
+        if func_abi['name'] == func_name:
+            return func_abi
+    raise ValueError('ABI of function {} is not found in vmc'.format(func_name))
+
+def mk_contract_tx_obj(
+        func_name,
+        args,
+        contract_addr,
+        contract_abi,
+        sender_addr,
+        value,
+        gas,
+        gas_price):
+    func_abi = get_func_abi(func_name, contract_abi)
+    arg_types = [arg_abi['type'] for arg_abi in func_abi['inputs']]
+    func_selector = eth_utils.function_abi_to_4byte_selector(func_abi)
+    data = func_selector + encode_abi(arg_types, args)
+    data = eth_utils.encode_hex(data)
```

Comment:
I would suggest using the `encode_hex` function from `evm.utils.hexadecimal.encode_hex`.  The `eth_utils.encode_hex` function may suffer a backwards incompatible breaking change in the coming months.

Reply:
No problem.